cou/steps/execute.py

```diff
@@ -89,25 +88,45 @@ async def _run_step(step: UpgradeStep, interactive: bool) -> None:
             await apply_plan(sub_step, interactive)
 
 
-async def apply_plan(plan: UpgradeStep, interactive: bool) -> None:
+async def apply_plan(plan: BaseStep, interactive: bool) -> None:
     """Apply the plan for upgrade.
 
     :param plan: Plan to be executed on steps.
-    :type plan: UpgradeStep
+    :type plan: BaseStep
     :param interactive:
     :type interactive: bool
     """
+    description = plan.description + " "
     result = ""
+    indicate_progress = True
+
+    # do nothing for empty upgrade step
+    if not plan:
+        return
+
+    # for UpgradePlan and its children class, do not show progress indicator
+    if isinstance(plan, UpgradePlan):
+        indicate_progress = False
+
+    # group and print all sub-steps with hierarchy for ApplicationUpgradePlan
+    if isinstance(plan, ApplicationUpgradePlan):
```

Comment:
If I understand it correctly, output of this will look like this:
```bash
Upgrade plan for 'rabbitmq-server' to victoria
    Upgrade software packages of 'rabbitmq-server' from the current APT repositories
    Upgrade 'rabbitmq-server' to the new channel: '3.9/stable'
    Change charm config of 'rabbitmq-server' 'source' to 'cloud:focal-victoria'
    Wait 300 s for model openstack to reach the idle state.
    Check if the workload of 'rabbitmq-server' has been upgraded
Continue(y/N)
Upgrade software packages of 'rabbitmq-server' from the current APT repositories ✔
Upgrade 'rabbitmq-server' to the new channel: '3.9/stable' ✔
...
```

Do you think printing full description in prompt is needed, since we will have indicator printing description of each sub-step.

Reply:
Yes I think it is necessary to let the user know what will be performed before they say `yes` to it.

Reply:
Like this UX will be to complicated and I think it's really not necessary, because the whole plan will be print before applying. If you can share a full example in PR description we can discuss it again.